### PR TITLE
Site health check for source tag.

### PR DESCRIPTION
**Pull request by zutigrm**
## Summary

Addresses issue:

- #5695 

## Relevant technical choices
* I moved the functionality to the new class instead of expanding on `Debug_Data` one

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: includes/Core/Util/Site_Health_Status.php [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+					$module->slug,
+					array(
+						Analytics_4::MODULE_SLUG,
+						AdSense::MODULE_SLUG,
+						Tag_Manager::MODULE_SLUG,
+					),
+					true
+				);
+			}
+		);
```
Comment (eugene-manuilov) [inline includes/Core/Util/Site_Health_Status.php:215]: The core class should know nothing about modules, it should only operate with entities available in the Core namespace. I think we need to rework your approach a bit:

1. Add a new interface `Module_With_Tag` to the `Core/Modules` namespace that has two methods: `register_tag()` and `has_tag( $content )`.
1. Add the new `Module_With_Tag` interface to `AdSense`, `Analytics_4` and `Tag_Manager` module classes.
1. Implement a new `has_tag( $content )` method in those classes that creates an appropriate tag matcher instance and return the result of the `has_tag` method call.
1. In this `get_active_modules ` method return only those modules that implement the `Module_With_Tag` interface.
1. Update the `check_if_tag_exists` method to accept a module instance instead of a module slug and call `$module->has_tag( ... )` method to determine whether the tag is added for that module.

File: includes/Core/Util/Site_Health_Status.php [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+		/**
+		 * Filters the search string used to check if a specific Google service tag exists in the content.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param string $search_string The initial search string constructed for tag detection.
+		 * @param string $module_slug   The slug of the module whose tag is being checked.
+		 * @param string $content       The content in which the tag is being searched for.
+		 */
+		$search_string = apply_filters( 'googlesitekit_site_status_check_if_tag_exists', $search_string, $module_slug, $content );
```
Comment (eugene-manuilov) [inline includes/Core/Util/Site_Health_Status.php:260]: This filter is also not needed. Let's remove it too. We rarely add new filters, only if they are really needed.

File: includes/Core/Util/Site_Health_Status.php [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+		if ( strpos( $content, $search_string ) !== false ) {
```
Comment (eugene-manuilov) [inline includes/Core/Util/Site_Health_Status.php:262]: Once the `has_tag` method is implemented in modules, we should probably better move this check into that method and update it to return a result that indicates whether there is a tag, a tag with comments or there is no tag at all. 

For this, we can add three constants to the `Module_Tag_Matchers` class:

```php
class Module_Tag_Matchers {

    const NO_TAG_FOUND = 0;
    const TAG_EXISTS = 1;
    const TAG_EXISTS_WITH_COMMENTS = 2;

}
```

And the use these constants to return from `has_tag` methods in modules.

File: includes/Core/Util/Site_Health_Status.php [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+use WP_REST_Server;
+
+/**
+ * Class for integrating status tab information with Site Health.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class Site_Health_Status {
```
Comment (eugene-manuilov) [inline includes/Core/Util/Site_Health_Status.php:31]: This is a bit confusing to have multiple classes for debug information, but splitting one giant class also makes sense. Let's probably create a new namespace `Core\Site_Health` and move `Debug_Data` and `Site_Health_Status` classes into it with the following changes:

1. Create a base class `Site_Health` that instantiates and registers other classes.
1. Move the `Debug_Data` class and rename it to be `General_Data`.
1. Move the `Site_Health_Status` class and rename it to be `Tags_Placement`.
1. Update the `Plugin` class to instantiate and register only the main `Site_Health` class.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @zutigrm. This is a good start, but there are things that we can make better. Please, see my comments for more details.

File: includes/Core/Modules/Tags/Module_Tag_Matchers.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		$tag_matchers = $this->regex_matchers();
+
+		$matches = array_filter(
+			$tag_matchers,
+			function( $pattern ) use ( $content ) {
+				return preg_match( $pattern, $content );
+			}
+		);
+
+		return ! empty( $matches );
```
[the remark this answers is not in the document]
Comment (zutigrm) [reply includes/Core/Modules/Tags/Module_Tag_Matchers.php:53]: I got carried away. Thanks, will update it

File: includes/Core/Util/Site_Health_Status.php [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+		if ( version_compare( $wp_version, '5.6', '<' ) ) {
+			$result['description'] = sprintf(
+				'<p>%s</p>',
+				__( 'This feature requires WordPress version 5.6 or higher', 'google-site-kit' )
+			);
+
+			return $result;
+		}
+
+		$response = wp_remote_get( site_url() ); // phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.wp_remote_get_wp_remote_get
```
[the remark this answers is not in the document]
Comment (zutigrm) [reply includes/Core/Util/Site_Health_Status.php:150]: That's great idea. Thanks

**Comment by zutigrm** [comment]
@eugene-manuilov Thank you for valuable feedback and great suggestions! I Updated the PR, with a slight adjustment to the originally suggested interface, instead of re-using the method, I made it as a trait that is easy to use in any module that needs to check for tags.

Let me know what you think

File: includes/Core/Util/Site_Health_Status.php (hunk 1 above)
Comment (zutigrm) [reply includes/Core/Util/Site_Health_Status.php]: I expanded on this approach, and made `Module_With_Tag_Trait` to avoid repeating the code and make it more centralised

File: includes/Core/Modules/Module_With_Tag.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+	/**
+	 * Checks if the module tag is found in the provided content.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $content Content to search for the tags.
+	 * @return bool TRUE if tag is found, FALSE if not.
+	 */
+	public function has_tag( $content );
```
Comment (eugene-manuilov) [inline includes/Core/Modules/Module_With_Tag.php:34]: Let's make this method to be named more specifically because `has_tag` is not very clear what it means. Maybe let's rename it to be `has_placed_tag_on_frontend`?

File: includes/Core/Modules/Module_With_Tag_Trait.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+		}
+		if ( 'analytics-4' === $this->slug ) {
+			$module_name = 'Analytics';
+		}
+		if ( 'tagmanager' === $this->slug ) {
+			$module_name = 'Tag Manager';
+		}
+
+		return $module_name;
+	}
```
Comment (eugene-manuilov) [inline includes/Core/Modules/Module_With_Tag_Trait.php:63]: If we can't make this trait be module agnostic, we should not use it at all because there is no point in it if we can't added it to another module without changing something within this trait to properly work with that module.

Let's get rid of this trait and add corresponding `has_tag` (or renamed version of that method) methods to each module that needs it. After doing it, we can drop another excessive method `get_tag_matchers` from all involved modules.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @zutigrm. This looks better now. There are still some comments we need to address though. Please, take a look at them.

File: includes/Core/Site_Health/Site_Health.php [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+	/**
+	 * Registers functionality through WordPress hooks.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function register() {
+		( new General_Data( $this->context, $this->options, $this->user_options, $this->authentication, $this->modules, $this->permissions ) )->register();
+		( new Tags_Placement( $this->modules ) )->register();
```
[the remark this answers is not in the document]
Comment (zutigrm) [reply includes/Core/Site_Health/Site_Health.php:112]: Thanks, I though to replicate what we are doing in `Plugin` class since this one is also just a main entry for the site health classes. I update the approach

File: includes/Core/Modules/Module_With_Tag.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+	/**
+	 * Checks if the module tag is found in the provided content.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $content Content to search for the tags.
+	 * @return bool TRUE if tag is found, FALSE if not.
+	 */
+	public function has_placed_tag_on_frontend( $content );
```
Comment (eugene-manuilov) [inline includes/Core/Modules/Module_With_Tag.php:32]: Oh... just realized that this method doesn't actually checks the frontend 🤦‍♂️. My bad, but could you please rename it to be `has_placed_tag_in_content` then? Sorry about this.

File: includes/Core/Site_Health/Site_Health.php [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+		Authentication $authentication,
+		Modules $modules,
+		Permissions $permissions
+	) {
+		$this->context        = $context;
+		$this->options        = $options;
+		$this->user_options   = $user_options;
+		$this->authentication = $authentication;
+		$this->modules        = $modules;
+		$this->permissions    = $permissions;
```
Comment (eugene-manuilov) [inline includes/Core/Site_Health/Site_Health.php:118]: No need to assign this variables in this class, we don't use it anywhere, it is just need to create new instances in this constructor.

File: includes/Core/Site_Health/Tags_Placement.php [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+		}
+		if ( 'analytics-4' === $slug ) {
+			$module_name = 'Analytics';
+		}
+		if ( 'tagmanager' === $slug ) {
+			$module_name = 'Tag Manager';
+		}
+
+		return $module_name;
+	}
```
Comment (eugene-manuilov) [inline includes/Core/Site_Health/Tags_Placement.php:277]: This method shouldn't be here. Let's probably add a new public abstract method `get_public_name` to the parent Module class that should be implemented in all children modules and then use that method in this class to get the public name of the module.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @zutigrm. This almost good to go. Just final comments from me and then I'll merge it.

File: includes/Core/Site_Health/Site_Health.php [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+		$context        = $context;
+		$options        = $options;
+		$user_options   = $user_options;
+		$authentication = $authentication;
+		$modules        = $modules;
+		$permissions    = $permissions;
+
+		$this->general_data   = new General_Data( $context, $options, $user_options, $authentication, $modules, $permissions );
+		$this->tags_placement = new Tags_Placement( $modules );
+	}
```
Comment (eugene-manuilov) [inline includes/Core/Site_Health/Site_Health.php:74]: I don't think we need these assignments, right? Let's remove them:
```suggestion
	) {
		$this->general_data   = new General_Data( $context, $options, $user_options, $authentication, $modules, $permissions );
		$this->tags_placement = new Tags_Placement( $modules );
	}
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @zutigrm. Looks good to me. There is just one thing that we need to remove before merging this PR. I'll apply that suggestion by myself.

**Review by eugene-manuilov — APPROVED** [review]

File: includes/Core/Site_Health/Site_Health.php (hunk 13 above)
Comment (zutigrm) [reply includes/Core/Site_Health/Site_Health.php]: Oh 🤦 Totally left it for no reason

File: includes/Core/Modules/Module.php [hunk 14]
```diff
@@ -155,6 +155,15 @@ public function __construct(
...
 	abstract public function register();
 
+	/**
+	 * Gets Module public name.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string Formatted module name.
+	 */
+	abstract public function get_public_name();
```
Comment (aaemnnosttv) [inline includes/Core/Modules/Module.php:165]: Why was this added? The name is already publicly accessible along with other data from `info` dynamically https://github.com/google/site-kit-wp/blob/d952edb61a7604dede836bc3df8c8b8eddc2afbd/includes/Core/Modules/Module.php#L42

File: includes/Core/Site_Health/General_Data.php [hunk 15]
```diff
@@ -28,7 +29,8 @@
  * @access private
  * @ignore
  */
-class Debug_Data {
```
Comment (aaemnnosttv) [inline includes/Core/Site_Health/General_Data.php:31]: Why rename the class? It was initially named this way for its relation to `\WP_Debug_Data::debug_data` which is what it extends. I'd prefer to keep it as before unless there is a reason it needs to change?

File: includes/Core/Site_Health/Tags_Placement.php [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+	 */
+	public function register() {
+		add_filter(
+			'googlesitekit_rest_routes',
+			function ( $rest_routes ) {
+				$health_check_routes = $this->get_rest_routes();
+
+				return array_merge( $rest_routes, $health_check_routes );
+			}
+		);
```
Comment (aaemnnosttv) [inline includes/Core/Site_Health/Tags_Placement.php:64]: Classes that extend our REST routes should be created as controllers that only implement the routes they add. This class also extends status tests which should be separate.

File: includes/Core/Site_Health/Tags_Placement.php [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+						'label' => __( 'Tags Placement', 'google-site-kit' ),
+						'test'  => $this->get_method_proxy( 'tags_placement_test' ),
+					);
+
+					return $tests;
+				}
+
+				$tests['async']['tag_placement'] = array(
+					'label'             => __( 'Tags Placement', 'google-site-kit' ),
+					'test'              => rest_url( 'google-site-kit/v1/core/site/data/tags-placement-test' ),
```
Comment (aaemnnosttv) [inline includes/Core/Site_Health/Tags_Placement.php:82]: We should use `\Google\Site_Kit\Core\REST_API\REST_Routes::REST_ROOT` in place of the hard coded `google-site-kit/v1` segment.

File: includes/Core/Site_Health/Tags_Placement.php [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+	private function get_rest_routes() {
+		return array(
+			new REST_Route(
+				'core/site/data/tags-placement-test',
+				array(
+					array(
+						'methods'             => WP_REST_Server::READABLE,
+						'callback'            => $this->get_method_proxy( 'tags_placement_test' ),
+						'permission_callback' => function () {
+							return current_user_can( Permissions::SETUP );
```
Comment (aaemnnosttv) [inline includes/Core/Site_Health/Tags_Placement.php:108]: This route should be available to anyone who can view Site Health.

File: includes/Core/Site_Health/Tags_Placement.php [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array Site health status results.
+	 */
+	protected function tags_placement_test() {
+		global $wp_version;
+
+		$result = array(
+			'label'   => __( 'Tags Placement', 'google-site-kit' ),
```
Comment (aaemnnosttv) [inline includes/Core/Site_Health/Tags_Placement.php:127]: This reads a bit oddly. Let's reword it to "Tag Placement".

File: includes/Core/Site_Health/Tags_Placement.php [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+			'status'  => 'recommended',
```
Comment (aaemnnosttv) [inline includes/Core/Site_Health/Tags_Placement.php:128]: This status shouldn't be recommended by default. See https://github.com/google/site-kit-wp/issues/5695#issuecomment-1896322675

File: includes/Core/Site_Health/Tags_Placement.php [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+
+		$active_modules = $this->get_active_modules();
+		if ( empty( $active_modules ) ) {
+			$result['description'] = sprintf(
+				'<p>%s</p>',
+				__( 'Tag status not available: AdSense, Tag Manager, and Analytics modules are not connected.', 'google-site-kit' )
+			);
+
+			return $result;
+		}
```
Comment (aaemnnosttv) [inline includes/Core/Site_Health/Tags_Placement.php:154]: This message should be more generic or not be hard coded with module names. We probably don't need to make the message dynamic, but it's also inaccurate right now. It's not possible to have an empty list of active modules – Search Console will always be active, as will Site Verification – they're `force_active`. Instead we can get the list of active modules that implement the interface and then check if they're connected. If they're all not connected, then we can say something like "no modules that place tags are connected."

File: includes/Core/Site_Health/Tags_Placement.php [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+
+		// Generate random page name that will result in 404 page, to prevent receiving
+		// cached page and target page with smaller content.
+		$random_page = substr(
+			str_shuffle( '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ' ),
+			0,
+			10
+		);
+		$response    = wp_remote_get( site_url( $random_page ) ); // phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.wp_remote_get_wp_remote_get
```
Comment (aaemnnosttv) [inline includes/Core/Site_Health/Tags_Placement.php:163]: This is more complicated than necessary and introduces a few problems as noted in https://github.com/google/site-kit-wp/issues/5695#issuecomment-1896322675

The main problem is that AdSense tags are intentionally not output on 404 pages, so that check is essentially broken. Second, `site_url` is about the base URL for where `wp-admin` is, not necessarily the front end. Instead, we should check `home_url` and use an added query parameter to serve as a cache buster, just like we do on the client https://github.com/google/site-kit-wp/blob/d952edb61a7604dede836bc3df8c8b8eddc2afbd/assets/js/googlesitekit/datastore/site/html.js#L52-L57

_Note: we should not include the `tagverify` parameter, as this suppresses all SK placed tags_

File: includes/Core/Site_Health/Tags_Placement.php [hunk 23]
```diff
@@ -0,0 +1,257 @@
...
+
+	/**
+	 * Gets active modules filtered to account only for
+	 * Analytics, AdSense and Tag Manager.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array Filtered active modules instances.
+	 */
+	protected function get_active_modules() {
```
Comment (aaemnnosttv) [inline includes/Core/Site_Health/Tags_Placement.php:200]: The name of the method here is misleading. It would be better if it conveyed that it returns active modules with tags, otherwise this is easy to misinterpret elsewhere where it's used.

File: includes/Core/Site_Health/Tags_Placement.php [hunk 24]
```diff
@@ -0,0 +1,257 @@
...
+		$active_modules = $this->get_active_modules();
+		if ( empty( $active_modules ) ) {
+			$result['description'] = sprintf(
+				'<p>%s</p>',
+				__( 'Tag status not available: AdSense, Tag Manager, and Analytics modules are not connected.', 'google-site-kit' )
+			);
+
+			return $result;
+		}
+
```
Comment (aaemnnosttv) [inline includes/Core/Site_Health/Tags_Placement.php:155]: Note there are other reasons why tags may not be output which would be normal and expected. i.e. https://github.com/google/site-kit-wp/blob/d952edb61a7604dede836bc3df8c8b8eddc2afbd/includes/Core/Tags/Guards/Tag_Environment_Type_Guard.php

or `useSnippet` is turned off.

File: includes/Modules/AdSense.php [hunk 25]
```diff
@@ -145,6 +147,17 @@ function () {
 		add_action( 'template_redirect', $this->get_method_proxy( 'register_tag' ) );
```
Comment (aaemnnosttv) [inline includes/Modules/AdSense.php:147]: The method proxy is no longer needed since it's now a public method.

File: includes/Modules/AdSense.php [hunk 26]
```diff
@@ -835,6 +849,32 @@ private function register_tag() {
...
+		} else {
+			foreach ( $tag_matchers as $pattern ) {
+				if ( preg_match( $pattern, $content ) ) {
+					return Module_Tag_Matchers::TAG_EXISTS;
+				}
+			}
+		}
+
+		return Module_Tag_Matchers::NO_TAG_FOUND;
+	}
```
Comment (aaemnnosttv) [inline includes/Modules/AdSense.php:876]: This seems like it would be the same in all classes and so we could extract it to a trait. We'd only need the class to have a method to provide the `Tag_Matchers` instance.

As mentioned in the issue though, using AMP removes all the comments so that would never match.

File: includes/Modules/AdSense.php [hunk 27]
```diff
@@ -835,6 +849,32 @@ private function register_tag() {
...
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $content Content to search for the tags.
+	 * @return bool TRUE if tag is found, FALSE if not.
+	 */
+	public function has_placed_tag_in_content( $content ) {
+		$tag_matchers = ( new Tag_Matchers() )->regex_matchers();
+
+		$search_string = 'Google AdSense snippet added by Site Kit';
```
Comment (aaemnnosttv) [inline includes/Modules/AdSense.php:863]: It's important to note that this string is localized when printed on the front end, so if the site had a non-EN language, this would fail to match. It might be good to search for the untranslated and translated versions of the string to improve the chances of finding it.

**Comment by aaemnnosttv** [comment]
Thanks @zutigrm – a few comments for you to consider regarding the follow-up we need on this one.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @zutigrm – a few comments for you to consider regarding the follow-up we need on this one.

File: includes/Core/Modules/Module.php (hunk 14 above)
Comment (eugene-manuilov) [reply includes/Core/Modules/Module.php:165]: Ugh... you are right, my bad, I forgot we have it already. So, we can probably replace that `get_public_name` method with just `$module->name` calls.

File: includes/Core/Site_Health/General_Data.php (hunk 15 above)
Comment (eugene-manuilov) [reply includes/Core/Site_Health/General_Data.php:31]: > It was initially named this way for its relation to `\WP_Debug_Data::debug_data` which is

I think this is something that only you remember in our team :). I didn't know it before.

> I'd prefer to keep it as before unless there is a reason it needs to change?

Do we really need to bind our class name to the `WP_Debug_Data` class? IMO "debug data" is something more technical and needed for troubleshooting, whilst "site health info" is something that explains the current status of the site.

Comment (zutigrm) [reply includes/Core/Site_Health/General_Data.php:31]: @aaemnnosttv The idea was to group site health checks/data, you find more info in this CR [comment](https://github.com/google/site-kit-wp/pull/8069#discussion_r1448871842)

cc: @eugene-manuilov

File: includes/Core/Site_Health/Tags_Placement.php (hunk 16 above)
Comment (zutigrm) [reply includes/Core/Site_Health/Tags_Placement.php:64]: Thank you pointing me to that

File: includes/Modules/AdSense.php (hunk 26 above)
Comment (zutigrm) [reply includes/Modules/AdSense.php:876]: It was moved between Trait and required method during CR. Added it as Trait, thanks
